diff --git a/src/amd/vulkan/nir/radv_nir_rt_stage_monolithic.c b/src/amd/vulkan/nir/radv_nir_rt_stage_monolithic.c
index 98ff3df2793b04418a0d28983dde0d25a1a0088b..0a472cd1adfa545d1b7ded57421ccd2afc3f75bb 100644
--- a/src/amd/vulkan/nir/radv_nir_rt_stage_monolithic.c
+++ b/src/amd/vulkan/nir/radv_nir_rt_stage_monolithic.c
@@ -278,6 +278,8 @@ radv_build_recursive_case(nir_builder *b, nir_def *idx, struct radv_ray_tracing_
    nir_inline_function_impl(b, nir_shader_get_entrypoint(shader), NULL, var_remap);
    nir_pop_if(b, NULL);
    ralloc_free(shader);
+
+   _mesa_hash_table_destroy(var_remap, NULL);
 }
 
 struct lower_rt_instruction_monolithic_state {
diff --git a/src/amd/vulkan/nir/radv_nir_rt_traversal_shader.c b/src/amd/vulkan/nir/radv_nir_rt_traversal_shader.c
index 2bc63601812fc746b4df64abc4b3e54eb53cc219..52088f200284425725bf7cb5a0fadf9fda80afc1 100644
--- a/src/amd/vulkan/nir/radv_nir_rt_traversal_shader.c
+++ b/src/amd/vulkan/nir/radv_nir_rt_traversal_shader.c
@@ -400,6 +400,8 @@ insert_inlined_shader(nir_builder *b, struct traversal_inlining_params *params,
    nir_push_if(b, nir_ieq_imm(b, idx, call_idx));
    nir_inline_function_impl(b, nir_shader_get_entrypoint(shader), NULL, var_remap);
    nir_pop_if(b, NULL);
+
+   _mesa_hash_table_destroy(var_remap, NULL);
 }
 
 static nir_function_impl *
diff --git a/src/amd/vulkan/radv_physical_device.c b/src/amd/vulkan/radv_physical_device.c
index 0bcd742bda88649d70e3b8411e7a89bea2933a4b..c104e4a17eb742f863f630236003e5ac33e15071 100644
--- a/src/amd/vulkan/radv_physical_device.c
+++ b/src/amd/vulkan/radv_physical_device.c
@@ -2827,6 +2827,7 @@ radv_physical_device_destroy(struct vk_physical_device *vk_device)
 
    radv_finish_wsi(pdev);
    ac_destroy_perfcounters(&pdev->ac_perfcounters);
+   free(pdev->perfcounters);
    if (pdev->addrlib)
       ac_addrlib_destroy(pdev->addrlib);
    if (pdev->ws)
diff --git a/src/amd/vulkan/radv_pipeline_rt.c b/src/amd/vulkan/radv_pipeline_rt.c
index 1a99632a06e4ee499e6505bd5f23380e5198669f..a079d8f2787f3f071376bd40758873a28bc7b153 100644
--- a/src/amd/vulkan/radv_pipeline_rt.c
+++ b/src/amd/vulkan/radv_pipeline_rt.c
@@ -883,6 +883,10 @@ radv_rt_compile_shaders(struct radv_device *device, struct vk_pipeline_cache *ca
          result = radv_rt_compile_nir(device, cache, pipeline, RADV_RT_LOWERING_MODE_FUNCTION_CALLS, &combined_stage,
                                       &payload_size, &hit_attrib_size, &stack_size, NULL, NULL, replay_block,
                                       skip_shaders_cache, has_position_fetch, &pipeline->groups[idx].ahit_isec_shader);
+         ralloc_free(final_shader);
+         if (isec && ahit)
+            ralloc_free(ahit);
+
          if (result != VK_SUCCESS)
             goto cleanup;
 
@@ -1062,6 +1066,7 @@ compile_rt_prolog(struct radv_device *device, struct radv_ray_tracing_pipeline *
    NIR_PASS(_, prolog_stage.nir, nir_opt_remove_phis);
 
    pipeline->prolog = radv_compile_rt_prolog(device, &prolog_stage, &debug);
+   ralloc_free(prolog_stage.nir);
 
    bool has_traversal = !!pipeline->base.base.shaders[MESA_SHADER_INTERSECTION];
 
